File: Plantemon/app/src/main/java/com/g4ng/logic/SwitchAction.java
package com.g4ng.logic;

import com.g4ng.model.Plant;
import com.g4ng.model.Player;

public class SwitchAction implements Action {
    private final Plant nextPlant;

    public SwitchAction(Plant nextPlant) {
        this.nextPlant = nextPlant;
    }

    @Override
    public void execute(Player performer, Player opponent, Action opponentAction) {
        if (nextPlant != null && !nextPlant.isDead()) {
            System.out.println(performer.getUsername() + " switched to " + nextPlant.getName() + "!");
            performer.setCurrentPlant(nextPlant);
        } else {
            System.out.println(performer.getUsername() + " failed to switch!");
        }
    }

    @Override
    public int getDefenseValue() {
        // Switching might leave you vulnerable (0 defense).
        return 0;
    }
}

File: plantemon-web/src/lib/domain/switch-action.ts
import type { Action } from "./types";
import type { Plant } from "./plant";
import type { Player } from "./player";

/** Port of logic/SwitchAction.java */
export class SwitchAction implements Action {
  private readonly nextPlant: Plant | null;

  constructor(nextPlant: Plant | null) {
    this.nextPlant = nextPlant;
  }

  execute(performer: Player, _opponent: Player, _opponentAction: Action | null): string {
    if (this.nextPlant !== null && !this.nextPlant.isDead()) {
      const result = `${performer.getUsername()} switched to ${this.nextPlant.getName()}!`;
      performer.setCurrentPlant(this.nextPlant);
      return result;
    }
    return `${performer.getUsername()} failed to switch!`;
  }

  getDefenseValue(): number {
    return 0;
  }
}
